File: welsch_adenet/competitors.py
"""Unified competitor solver and robust BIC tuning.

Supports various loss functions (squared, absolute/LAD, Huber, Tukey, Welsch)
with adaptive elastic-net penalty, solved via proximal Adam and tuned by Robust BIC.
"""

import numpy as np
from .init_scale import robust_init, adaptive_weights

# 1. Loss Functions and Derivatives (Psi)

def loss_squared(u):
    return 0.5 * u * u

def psi_squared(u):
    return u

def loss_absolute(u):
    return np.abs(u)

def psi_absolute(u):
    return np.sign(u)

def loss_huber(u, k=1.345):
    abs_u = np.abs(u)
    return np.where(abs_u <= k, 0.5 * u * u, k * abs_u - 0.5 * k * k)

def psi_huber(u, k=1.345):
    return np.where(np.abs(u) <= k, u, k * np.sign(u))

def loss_tukey(u, d=4.685):
    abs_u = np.abs(u)
    inside = (d**2 / 6.0) * (1.0 - (1.0 - (u / d) ** 2) ** 3)
    outside = d**2 / 6.0
    return np.where(abs_u <= d, inside, outside)

def psi_tukey(u, d=4.685):
    return np.where(np.abs(u) <= d, u * (1.0 - (u / d) ** 2) ** 2, 0.0)

def loss_welsch(u, c=2.11):
    return c * c * (1.0 - np.exp(-(u * u) / (2.0 * c * c)))

def psi_welsch(u, c=2.11):
    return u * np.exp(-(u * u) / (2.0 * c * c))


LOSS_REGISTRY = {
    "squared": (loss_squared, psi_squared),
    "absolute": (loss_absolute, psi_absolute),
    "huber": (loss_huber, psi_huber),
    "tukey": (loss_tukey, psi_tukey),
    "welsch": (loss_welsch, psi_welsch),
}

# 2. Unified Solver

def _soft_ridge_prox(v, step, l1, l2, w):
    thresh = step * l1 * w
    shrunk = np.sign(v) * np.maximum(np.abs(v) - thresh, 0.0)
    return shrunk / (1.0 + step * l2)

def fit_competitor(
    X, y, l1, l2, weights, sigma, loss_type="squared",
    eta=1e-2, omega1=0.9, omega2=0.999, eps=1e-8,
    beta0=None, max_iter=2000, tol=1e-6, loss_param=None
):
    """Fit a model with a given loss function and adaptive elastic-net penalty."""
    n, p = X.shape
    beta = np.zeros(p) if beta0 is None else beta0.astype(float).copy()
    m = np.zeros(p)
    v = np.zeros(p)

    loss_fn, psi_fn = LOSS_REGISTRY[loss_type]

    # Setup loss parameter (e.g. k, d, c)
    kwargs = {}
    if loss_param is not None:
        if loss_type == "huber":
            kwargs["k"] = loss_param
        elif loss_type == "tukey":
            kwargs["d"] = loss_param
        elif loss_type == "welsch":
            kwargs["c"] = loss_param

    r = y - X @ beta
    prev_loss = loss_fn(r / sigma, **kwargs)

    for t in range(1, max_iter + 1):
        # gradient of the smooth loss term (subgradient for absolute loss)
        g = -(X.T @ psi_fn(r / sigma, **kwargs)) / sigma

        # Adam moments + bias correction
        m = omega1 * m + (1.0 - omega1) * g
        v = omega2 * v + (1.0 - omega2) * (g * g)
        m_hat = m / (1.0 - omega1 ** t)
        v_hat = v / (1.0 - omega2 ** t)

        # per-coordinate step, descent point, prox
        step = eta / (np.sqrt(v_hat) + eps)
        vtent = beta - step * m_hat
        beta = _soft_ridge_prox(vtent, step, l1, l2, weights)

        # stop on average change in the loss
        r = y - X @ beta
        loss = loss_fn(r / sigma, **kwargs)
        if np.linalg.norm(loss - prev_loss) / n < tol:
            break
        prev_loss = loss

    return (1.0 + l2 / n) * beta

# 3. Tuning

def rbic_score(X, y, beta, sigma, loss_type, loss_param=None):
    n = X.shape[0]
    r = y - X @ beta
    loss_fn, _ = LOSS_REGISTRY[loss_type]
    
    kwargs = {}
    if loss_param is not None:
        if loss_type == "huber":
            kwargs["k"] = loss_param
        elif loss_type == "tukey":
            kwargs["d"] = loss_param
        elif loss_type == "welsch":
            kwargs["c"] = loss_param
            
    deviance = loss_fn(r / sigma, **kwargs).sum()
    active = int(np.sum(np.abs(beta) > 1e-8))
    return deviance + np.log(n) * active

def tune_competitor(
    X, y, loss_type="squared", loss_param=None, gamma=1.0, 
    n_l1=20, n_l2=5, l1_grid=None, l2_grid=None, **adam_kw
):
    """Grid search over (l1, l2) minimizing robust BIC."""
    n, p = X.shape
    beta0, sigma = robust_init(X, y)
    
    # LAD-Lasso is usually non-adaptive (we can set weights to 1s)
    # But for other methods we use adaptive weights
    if loss_type == "absolute":
        w = np.ones(p)  # non-adaptive L1 penalty for standard LAD-Lasso
    else:
        w = adaptive_weights(beta0, gamma=gamma)

    if l1_grid is None or l2_grid is None:
        # data-driven lambda_1 max
        l1_max = np.max(np.abs(X.T @ (y / sigma))) / n
        l1_grid = np.logspace(np.log10(l1_max), np.log10(l1_max * 1e-3), n_l1)
        l2_grid = np.concatenate([[0.0], np.logspace(-3, 0, n_l2 - 1)])

    best = {"rbic": np.inf}
    for l2 in l2_grid:
        warm = beta0.copy()
        for l1 in l1_grid:
            beta = fit_competitor(
                X, y, l1, l2, w, sigma, loss_type=loss_type, 
                loss_param=loss_param, beta0=warm, **adam_kw
            )
            warm = beta
            score = rbic_score(X, y, beta, sigma, loss_type, loss_param)
            if score < best["rbic"]:
                best = {
                    "beta": beta, "lambda1": l1, "lambda2": l2,
                    "sigma": sigma, "weights": w, "rbic": score
                }
    return best

File: r-package/welschAdEnet/R/competitors.R
loss_squared <- function(u) 0.5 * u * u
psi_squared <- function(u) u

loss_absolute <- function(u) abs(u)
psi_absolute <- function(u) sign(u)

loss_huber <- function(u, k = 1.345) {
  abs_u <- abs(u)
  ifelse(abs_u <= k, 0.5 * u * u, k * abs_u - 0.5 * k * k)
}
psi_huber <- function(u, k = 1.345) {
  ifelse(abs(u) <= k, u, k * sign(u))
}

loss_tukey <- function(u, d = 4.685) {
  abs_u <- abs(u)
  inside <- (d^2 / 6) * (1 - (1 - (u / d)^2)^3)
  outside <- d^2 / 6
  ifelse(abs_u <= d, inside, outside)
}
psi_tukey <- function(u, d = 4.685) {
  ifelse(abs(u) <= d, u * (1 - (u / d)^2)^2, 0)
}

loss_welsch_ <- function(u, c = 2.11) welsch_loss(u, c)
psi_welsch_ <- function(u, c = 2.11) welsch_psi(u, c)

.loss_registry <- list(
  squared  = list(loss = loss_squared,  psi = psi_squared),
  absolute = list(loss = loss_absolute, psi = psi_absolute),
  huber    = list(loss = loss_huber,    psi = psi_huber),
  tukey    = list(loss = loss_tukey,    psi = psi_tukey),
  welsch   = list(loss = loss_welsch_,  psi = psi_welsch_)
)

loss_param_args <- function(loss_type, loss_param) {
  if (is.null(loss_param)) return(list())
  switch(loss_type,
    huber  = list(k = loss_param),
    tukey  = list(d = loss_param),
    welsch = list(c = loss_param),
    list()
  )
}

#' Fit a penalized robust-regression competitor
#'
#' A single proximal Adam solve generalizing [welsch_adenet()] to a
#' selectable robust loss, so that alternative estimators can be benchmarked
#' against the Welsch adaptive elastic-net with the same solver and adaptive
#' elastic-net penalty.
#'
#' @param X Numeric design matrix, `n` by `p`.
#' @param y Numeric response vector, length `n`.
#' @param l1 Non-negative scalar, the adaptive lasso penalty \eqn{\lambda_1}.
#' @param l2 Non-negative scalar, the ridge penalty \eqn{\lambda_2}.
#' @param weights Numeric vector of length `p`, the adaptive lasso weights.
#' @param sigma Positive scalar, the robust scale estimate.
#' @param loss_type One of `"squared"`, `"absolute"`, `"huber"`, `"tukey"`,
#'   or `"welsch"`.
#' @param eta Numeric scalar, the Adam base step size.
#' @param omega1,omega2 Numeric scalars in `[0, 1)`, the Adam first- and
#'   second-moment decay rates.
#' @param eps Small positive constant preventing division by zero in the
#'   Adam step.
#' @param beta0 Optional numeric warm start of length `p` (defaults to a
#'   zero vector).
#' @param max_iter Maximum number of proximal Adam iterations.
#' @param tol Relative loss-change convergence tolerance.
#' @param loss_param Optional tuning constant for the chosen loss (`k` for
#'   Huber, `d` for Tukey, `c` for Welsch); `NULL` uses each loss's usual
#'   default constant.
#' @return Numeric vector of length `p`, the rescaled coefficient estimate.
#' @seealso [tune_competitor()] to also select `(l1, l2)` by robust BIC,
#'   [welsch_adenet()] for the Welsch-specific solver this generalizes.
#' @examples
#' set.seed(1)
#' X <- matrix(rnorm(80 * 4), 80, 4)
#' y <- as.numeric(X %*% c(1, -1, 0, 0)) + rnorm(80)
#' init <- robust_init(X, y)
#' w <- adaptive_weights(init$beta0)
#' beta_hat <- fit_competitor(X, y, l1 = 0.05, l2 = 0.01, weights = w,
#'                             sigma = init$sigma, loss_type = "huber")
#' round(beta_hat, 2)
#' @export
fit_competitor <- function(X, y, l1, l2, weights, sigma, loss_type = "squared",
                            eta = 1e-2, omega1 = 0.9, omega2 = 0.999, eps = 1e-8,
                            beta0 = NULL, max_iter = 2000L, tol = 1e-6,
                            loss_param = NULL) {
  n <- nrow(X)
  p <- ncol(X)
  beta <- if (is.null(beta0)) numeric(p) else as.numeric(beta0)
  m <- numeric(p)
  v <- numeric(p)

  reg <- .loss_registry[[loss_type]]
  if (is.null(reg)) stop("unknown loss_type: ", loss_type)
  loss_args <- loss_param_args(loss_type, loss_param)

  r <- y - as.numeric(X %*% beta)
  prev_loss <- do.call(reg$loss, c(list(r / sigma), loss_args))

  for (t in seq_len(max_iter)) {
    # gradient of the smooth loss term (subgradient for the absolute loss)
    psi_val <- do.call(reg$psi, c(list(r / sigma), loss_args))
    g <- -as.numeric(crossprod(X, psi_val)) / sigma

    # Adam moments + bias correction
    m <- omega1 * m + (1 - omega1) * g
    v <- omega2 * v + (1 - omega2) * (g * g)
    m_hat <- m / (1 - omega1^t)
    v_hat <- v / (1 - omega2^t)

    # per-coordinate step, descent point, prox
    step <- eta / (sqrt(v_hat) + eps)
    vtent <- beta - step * m_hat
    beta <- soft_ridge_prox(vtent, step, l1, l2, weights)

    # stop on average change in the loss
    r <- y - as.numeric(X %*% beta)
    loss <- do.call(reg$loss, c(list(r / sigma), loss_args))
    if (sqrt(sum((loss - prev_loss)^2)) / n < tol) break
    prev_loss <- loss
  }

  (1 + l2 / n) * beta
}

competitor_rbic_score <- function(X, y, beta, sigma, loss_type, loss_param = NULL) {
  n <- nrow(X)
  r <- y - as.numeric(X %*% beta)
  reg <- .loss_registry[[loss_type]]
  loss_args <- loss_param_args(loss_type, loss_param)
  deviance <- sum(do.call(reg$loss, c(list(r / sigma), loss_args)))
  active <- sum(abs(beta) > 1e-8)
  deviance + log(n) * active
}

#' Tune a competitor loss by grid search over robust BIC
#'
#' Grid search over `(lambda1, lambda2)` for a given `loss_type`, minimizing
#' a robust-BIC-style score (the summed loss over standardized residuals
#' plus a BIC complexity penalty on the active set), analogous to
#' [fit_rbic()] but for the competing losses in [fit_competitor()].
#'
#' @param X Numeric design matrix, `n` by `p`.
#' @param y Numeric response vector, length `n`.
#' @param loss_type One of `"squared"`, `"absolute"`, `"huber"`, `"tukey"`,
#'   or `"welsch"`.
#' @param loss_param Optional tuning constant for the chosen loss; see
#'   [fit_competitor()].
#' @param gamma Adaptive-weight exponent. Ignored for
#'   `loss_type = "absolute"`, which uses uniform weights (matching the
#'   usual non-adaptive LAD-lasso).
#' @param n_l1,n_l2 Sizes of the default `(lambda1, lambda2)` grids, used
#'   only when `l1_grid`/`l2_grid` are `NULL`.
#' @param l1_grid,l2_grid Optional numeric penalty grids.
#' @param ... Additional arguments passed to [fit_competitor()].
#' @return A list with elements `beta`, `lambda1`, `lambda2`, `sigma`,
#'   `weights`, and `rbic` (the best score found).
#' @examples
#' \donttest{
#' set.seed(1)
#' X <- matrix(rnorm(80 * 4), 80, 4)
#' y <- as.numeric(X %*% c(1, -1, 0, 0)) + rnorm(80)
#' res <- tune_competitor(X, y, loss_type = "huber",
#'                         l1_grid = c(0.1, 0.02), l2_grid = c(0, 0.1))
#' round(res$beta, 2)
#' }
#' @export
tune_competitor <- function(X, y, loss_type = "squared", loss_param = NULL,
                             gamma = 1.0, n_l1 = 20L, n_l2 = 5L,
                             l1_grid = NULL, l2_grid = NULL, ...) {
  n <- nrow(X)
  p <- ncol(X)
  init <- robust_init(X, y)
  beta0 <- init$beta0
  sigma <- init$sigma

  # LAD-lasso is usually non-adaptive (uniform weights); other losses use
  # the adaptive weights.
  w <- if (identical(loss_type, "absolute")) rep(1, p) else adaptive_weights(beta0, gamma = gamma)

  if (is.null(l1_grid) || is.null(l2_grid)) {
    l1_max <- max(abs(crossprod(X, y / sigma))) / n
    l1_grid <- 10^seq(log10(l1_max), log10(l1_max * 1e-3), length.out = n_l1)
    l2_grid <- c(0, 10^seq(-3, 0, length.out = n_l2 - 1))
  }

  best <- list(rbic = Inf)
  for (l2 in l2_grid) {
    warm <- beta0
    for (l1 in l1_grid) {
      beta <- fit_competitor(X, y, l1, l2, w, sigma, loss_type = loss_type,
                              loss_param = loss_param, beta0 = warm, ...)
      warm <- beta
      score <- competitor_rbic_score(X, y, beta, sigma, loss_type, loss_param)
      if (score < best$rbic) {
        best <- list(beta = beta, lambda1 = l1, lambda2 = l2,
                      sigma = sigma, weights = w, rbic = score)
      }
    }
  }
  best
}
